bentoml/_internal/runner/runnable.py

```diff
@@ -24,19 +24,101 @@
 RUNNABLE_METHOD_MARK: str = "_bentoml_runnable_method"
 
 
-class Runnable(ABC):
-    SUPPORT_NVIDIA_GPU: bool
-    SUPPORT_CPU_MULTI_THREADING: bool
+class RunnableMeta(ABCMeta, t.Type[t.Any]):
+    supported_resources: Set[str]
+    supports_multi_threading: bool
 
     methods: dict[str, RunnableMethod[t.Any, t.Any, t.Any]] | None = None
 
-    @abstractmethod
-    def __init__(self, **kwargs: t.Any):
-        ...
+    def __new__(
+        cls,
+        name: str,
+        bases: t.Tuple[type, ...],
+        attr_dict: dict[t.Any, t.Any],
+        *,
+        supported_resources: Set[str] | None = None,
+        supports_multi_threading: bool | None = None,
+        **_kwargs: t.Any,
+    ) -> RunnableMeta:
+        res = super().__new__(cls, name, bases, attr_dict)
+
+        if "SUPPORT_NVIDIA_GPU" in attr_dict:
+            if supported_resources is None:
+                if attr_dict["SUPPORT_NVIDIA_GPU"]:
+                    supported_resources = {"nvidia.com/gpu"}
+                else:
+                    supported_resources = set()
+                logger.warning(
+                    f"{name} is using deprecated 'SUPPORT_NVIDIA_GPU'. Please convert to using 'supported_resources':\n"
+                    f"class {name}(Runnable, supported_resources={supported_resources}):\n"
+                    "    ..."
+                )
+            else:
+                logger.warning(
+                    f"Deprecated 'SUPPORT_NVIDIA_GPU' is being ignored in favor of 'supported_resources' for {name}."
+                )
+
+        if supported_resources is None:
+            # attempt to get supported_resources from a superclass
+            found_base = None
+            for base in bases:
```

Comment:
For simplicity, should we just assume `supported_resources` must be provided in the leaf runnable class, so we do not have to check for base classes?

Reply:
We actually do this ourselves in [`pytorch`](https://github.com/bentoml/BentoML/blob/21267df1fc4239d4a9ae3d0c682f4cbd6dbf94da/bentoml/_internal/frameworks/common/pytorch.py#L46), which is why I added this. We could just pass the parameters to the base class but it's a little ugly on the framework impl side.